### PR TITLE
Follow-up to #340: tighten workspace-to-complog support

### PR DESCRIPTION
**Pull request by slang25**
Hi @jaredpar — this is a continuation of #340 (which I've been reviewing). The original PR landed the core of `CompilerLogUtil.CreateFromWorkspace` / `CompilerLogBuilder.AddFromWorkspace`; this builds on top of it with a few tweaks I think are worth folding in before merging.

## What's in here

- **Non-throwing variant.** Adds `TryCreateFromWorkspace` returning `CreateFromWorkspaceResult { Succeeded, CompilerCalls, Diagnostics }`. The throwing `CreateFromWorkspace` is kept and now layers on top of it. `AddFromWorkspace` returns the synthesized `CompilerCall` (or `null` on failure) so callers can correlate diagnostics with projects.
- **TargetFramework recovery.** `MSBuildWorkspace` names multi-targeted projects `AssemblyName(tfm)`, so we parse that first; falls back to the parent directory of `OutputFilePath` for default-layout SDK projects (best-effort — `artifacts/` layouts give `{config}_{tfm}` and won't round-trip cleanly).
- **Source-generated documents.** `project.GetSourceGeneratedDocumentsAsync` results are now serialized as `RawContentKind.GeneratedText` and `IncludesGeneratedText = true`, so the reader uses the captured generator output rather than re-running generators against possibly-different analyzer versions.
- **Project references via the Solution.** Switched from `CompilationReference` to walking `project.ProjectReferences` and resolving through `parentProject.Solution.GetProject(...)`. Prefers the dep's on-disk `OutputFilePath` (which matches the consumer's TFM) and only emits the in-memory `Compilation` when nothing is on disk.
- **Crypto key file.** When `compilation.Options.CryptoKeyFile` is set and resolves to a real file (relative to the project directory if not rooted), it gets captured as `RawContentKind.CryptoKeyFile`.
- **Smaller polish.** `ChecksumAlgorithm` taken from the first document's `SourceText` rather than hardcoded SHA-256; `ParseOptions` read from `project.ParseOptions` (the authoritative source) rather than from a syntax tree; `.vbproj` extension fallback for VB projects with no `FilePath`; cancellation propagated through `TryCreateFromWorkspace`; the analyzer-skip diagnostic now suggests reloading with `BasicAnalyzerKind.OnDisk` / `None`.
- **Tests** for round-trip, source-text preservation, project-reference handling, the file-path overload, on-disk analyzer references, generated-text round-trip, cancellation, and a regression lock that the stored args stay empty.

## What I tried and backed out

I initially added a `WorkspaceCommandLineSynthesizer` that reconstructed a csc/vbc-style rsp from `compilation.Options` + `project.MetadataReferences` + analyzers (you can see it at b5161e0). The intent was to give `complog rsp` / `replay` / `export` something to work with for workspace-derived complogs.

I hadn't realised that the Roslyn workspace API genuinely doesn't surface emit-time inputs — embedded resources, Win32 manifest/icon/resource, source link, app.config, etc. live on `Compilation.Emit()` parameters and are supplied by the host (MSBuild), not on `Project`. So a synthesized rsp would always quietly drop those inputs and produce a binary that compiles but isn't faithful. That's worse than failing loudly, so I removed the synthesizer in 250bd2e and left the stored args empty — `replay`/`rsp`/`export` will fail visibly on workspace-derived complogs, which is the honest behaviour. There's a test that locks in the empty-args invariant.

If you'd rather take the original synthesizer-included version and document the partial fidelity instead, happy to revert that second commit.

**Update:** merged `main` in (clean, just a `using` clash) and pushed a few more fixes from another review pass over the branch:

- The emitted project-reference cache is now keyed by `ProjectId` rather than `"{AssemblyName}.dll"` — assembly names aren't unique across a workspace (each TargetFramework flavour of a multi-targeted project shares one), so the old key could hand one flavour's PE bytes to the other flavour's consumer. There's a regression test locking this in.
- The serialization core is now properly async (`AddFromWorkspaceAsync` awaits `GetCompilationAsync`/`GetTextAsync` instead of `.GetAwaiter().GetResult()`), with `CreateFromWorkspaceAsync`/`TryCreateFromWorkspaceAsync` exposed and the sync overloads kept as thin wrappers. If that's more public surface than you'd like, happy to trim to just one set — no strong opinion here from me.
- `AddAssembly` and the emitted-assembly path now share the zip entry write, ironing out some duplicated logic.
- The project-reference test no longer depends on the fixture's shape (it used to quietly skip when there were no P2P refs) — it now builds `AdhocWorkspace` projects directly, exercises the in-memory emit path, and asserts the consumer re-compiles without errors using only what's stored in the log.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/Basic.CompilerLog.Util/CompilerLogBuilder.cs [hunk 1]
```diff
@@ -228,6 +230,361 @@ void AddIf(string? filePath, bool isRefAssembly)
...
+            return null;
+        }
+
+        var allProjectReferencesAdded = true;
+
+        // Roslyn's Workspace API doesn't surface emit-time inputs (embedded resources,
+        // Win32 manifest/icon/resource, source link, app.config) — only items needed for
+        // semantic analysis. Synthesizing a partial command line from compilation.Options
+        // would produce an rsp that compiles but silently drops those inputs, so we leave
+        // the args empty: replay/rsp/export will fail visibly rather than misleadingly.
```
Comment (jaredpar) [inline src/Basic.CompilerLog.Util/CompilerLogBuilder.cs]: My lean is that command line should be synthesized with the understanding that it will be lossy / inaccurate in edge cases. If there is no command line then pretty much every command here (replay, export, etc ...) isn't useful on the logs and we'll need cases to test them out. It should be possible to synthesize the command line args that represent the `Compilation` we're given here, even if there are a few edges around emit. 

Could even note during replay / export that the logs originated from a workspace hence they may have fidelity issues. Already give some fidelity warnings for these commands.

File: src/Basic.CompilerLog.Util/CompilerLogBuilder.cs [hunk 2]
```diff
@@ -228,6 +230,361 @@ void AddIf(string? filePath, bool isRefAssembly)
...
+        var compilationDataPackHash = await CreateCompilationDataPackAsync().ConfigureAwait(false);
+        if (!allProjectReferencesAdded)
+        {
+            // A dropped project reference means the stored compilation would silently differ
+            // from the workspace one, so the project is excluded rather than recorded broken.
+            return null;
+        }
```
Comment (jaredpar) [inline src/Basic.CompilerLog.Util/CompilerLogBuilder.cs]: It took me a bit to understand that `allProjectReferencesAdded` was being changed within that local function. I would favor a tupled return from `CreateCompilationPackAsync` that had this as part of it.

File: src/Basic.CompilerLog.Util/CompilerLogBuilder.cs [hunk 3]
```diff
@@ -228,6 +230,361 @@ void AddIf(string? filePath, bool isRefAssembly)
...
+            var dataPack = new CompilationDataPack()
+            {
+                ContentList = new(),
+                ValueMap = new(),
+                References = new(),
+                Analyzers = new(),
+                Resources = new(),
+                ChecksumAlgorithm = firstSourceText?.ChecksumAlgorithm ?? SourceHashAlgorithm.Sha256,
+                EmitPdb = false,
+                HasGeneratedFilesInPdb = false,
```
Comment (jaredpar) [inline src/Basic.CompilerLog.Util/CompilerLogBuilder.cs]: Think that `null` is the right value cause we don't really know. At the same time this field really isn't used at the moment, seeing if I can remove it in another PR.

**Comment by slang25** [comment]
@jaredpar what do you think to this feature btw? I'm using this branch at work, and so far it's working for my purposes.

File: src/Basic.CompilerLog.UnitTests/CompilerLogBuilderTests.cs [hunk 4]
```diff
@@ -153,6 +157,276 @@ public void CompilerFilePathMissingCommitHash()
         });
     }
 
+    private Workspace LoadConsoleWorkspace(BasicAnalyzerKind analyzerKind = BasicAnalyzerKind.None)
```
Comment (jaredpar) [inline src/Basic.CompilerLog.UnitTests/CompilerLogBuilderTests.cs:160]: There should also be a test in `UsingAllCompilerLogsTests`. Essentially, for every log open the log using the `SolutionReader`, then export that to a new compiler log, read the `Compilation` out of that and verify there are no diagnostics. That gets the round tripping for all of the core scenarios that we've identified and means future scenarios get tested too. 

It's possible because of the lack of fidelity in going `Workspace` to compiler log, some resulting `Compilations` will have diagnostics. If so and they're unresolvable they can be excluded from that test.

File: src/Basic.CompilerLog.Util/CompilerLogBuilder.cs [hunk 5]
```diff
@@ -228,6 +230,361 @@ void AddIf(string? filePath, bool isRefAssembly)
...
+        var compilation = await project.GetCompilationAsync(cancellationToken).ConfigureAwait(false);
+        if (compilation is null)
+        {
+            Diagnostics.Add($"Cannot get compilation for {project.Name}");
+            return null;
+        }
+
+        var allProjectReferencesAdded = true;
+
+        // Roslyn's Workspace API doesn't surface emit-time inputs (embedded resources,
```
Comment (jaredpar) [inline src/Basic.CompilerLog.Util/CompilerLogBuilder.cs:259]: I checked with the team to see if say `MSBuildWorkspace` actually had the emit data available in an API we weren't aware of. They confirmed that it does not exist and they intentionally do not get any emit data in those scenarios.

File: src/Basic.CompilerLog.Util/CompilerLogBuilder.cs [hunk 6]
```diff
@@ -228,6 +230,361 @@ void AddIf(string? filePath, bool isRefAssembly)
...
+                : null;
+
+            dataPack.ValueMap["assemblyFileName"] = assemblyFileName;
+            dataPack.ValueMap["outputDirectory"] = outputDirectory;
+            dataPack.ValueMap["xmlFilePath"] = null;
+            dataPack.ValueMap["compilationName"] = project.AssemblyName;
+
+            foreach (var document in project.Documents)
+            {
+                var filePath = document.FilePath ?? document.Name;
```
Comment (jaredpar) [inline src/Basic.CompilerLog.Util/CompilerLogBuilder.cs]: This can leave `filePath` as relative, should we force it to be full by qualifying it off of the project directory? Believe that the existing code mostly forces qualified path particularly for source.

File: src/Basic.CompilerLog.Util/CompilerLogBuilder.cs [hunk 7]
```diff
@@ -228,6 +230,361 @@ void AddIf(string? filePath, bool isRefAssembly)
...
+            return WriteContentMessagePack(dataPack);
+        }
+
+        void AddWorkspaceCompilationOptions(CompilationInfoPack infoPack, Compilation compilation, bool isCSharp)
+        {
+            infoPack.EmitOptionsHash = WriteContentMessagePack(MessagePackUtil.CreateEmitOptionsPack(new EmitOptions()));
+
+            if (isCSharp)
+            {
+                var parseOptions = (project.ParseOptions as CSharpParseOptions) ?? CSharpParseOptions.Default;
```
Comment (jaredpar) [inline src/Basic.CompilerLog.Util/CompilerLogBuilder.cs]:
```suggestion
                var parseOptions = (CSharpParseOptions)(project.ParseOption);
```

File: src/Basic.CompilerLog.Util/CompilerLogBuilder.cs [hunk 8]
```diff
@@ -228,6 +230,361 @@ void AddIf(string? filePath, bool isRefAssembly)
...
+                infoPack.ParseOptionsHash = WriteContentMessagePack(
+                    MessagePackUtil.CreateCSharpParseOptionsPack(parseOptions));
+                infoPack.CompilationOptionsHash = WriteContentMessagePack(
+                    MessagePackUtil.CreateCSharpCompilationOptionsPack((CSharpCompilationOptions)compilation.Options));
+            }
+            else
+            {
+                var parseOptions = (project.ParseOptions as VisualBasicParseOptions) ?? VisualBasicParseOptions.Default;
```
Comment (jaredpar) [inline src/Basic.CompilerLog.Util/CompilerLogBuilder.cs]:
```suggestion
                var parseOptions = (VisualBasicParseOptions)(project.ParseOptions);
```

File: src/Basic.CompilerLog.Util/CompilerLogBuilder.cs [hunk 9]
```diff
@@ -228,6 +230,361 @@ void AddIf(string? filePath, bool isRefAssembly)
...
+                    MessagePackUtil.CreateVisualBasicParseOptionsPack(parseOptions));
+                infoPack.CompilationOptionsHash = WriteContentMessagePack(
+                    MessagePackUtil.CreateVisualBasicCompilationOptionsPack((VisualBasicCompilationOptions)compilation.Options));
+            }
+        }
+    }
+
+    private void AddSourceText(CompilationDataPack dataPack, RawContentKind kind, string filePath, SourceText sourceText)
+    {
+        var encoding = sourceText.Encoding ?? ContentEncoding;
```
Comment (jaredpar) [inline src/Basic.CompilerLog.Util/CompilerLogBuilder.cs:507]: Not sure this will round trip unless `sourceText.Encoding` has a BOM. The `SourceText.From` API that is called in the reader will effectively use UTF8 as a reader unless there is a detectable BOM.

File: src/Basic.CompilerLog.Util/CompilerLogBuilder.cs [hunk 10]
```diff
@@ -228,6 +230,361 @@ void AddIf(string? filePath, bool isRefAssembly)
...
+                FilePath = filePath,
+                AssemblyName = info.AssemblyName,
+                AssemblyInformationalVersion = info.AssemblyInformationalVersion,
+                NetModuleMvids = ImmutableArray<Guid>.Empty,
+            });
+    }
+
+    private BuilderAssemblyData AddEmittedAssembly(string displayPath, MemoryStream stream)
+    {
+        stream.Position = 0;
```
Comment (jaredpar) [inline src/Basic.CompilerLog.Util/CompilerLogBuilder.cs]: Would favor a `Debug.Assert(stream.Position == 0);` and at the end of this method make sure that we reset to `0` before returning.

File: src/Basic.CompilerLog.Util/CompilerLogBuilder.cs [hunk 11]
```diff
@@ -228,6 +230,361 @@ void AddIf(string? filePath, bool isRefAssembly)
...
+
+        var projectDir = project.FilePath is { } projectFilePath
+            ? Path.GetDirectoryName(projectFilePath)
+            : null;
+        return projectDir is not null
+            ? Path.Combine(projectDir, path)
+            : null;
+    }
+
+    private static string GetWorkspaceAssemblyFileName(string assemblyName, OutputKind outputKind) =>
```
Comment (jaredpar) [inline src/Basic.CompilerLog.Util/CompilerLogBuilder.cs]: Would make this a local function

File: src/Basic.CompilerLog.Util/CompilerLogUtil.cs [hunk 12]
```diff
@@ -29,6 +30,29 @@ public ConvertBinaryLogResult(bool succeeded, List<CompilerCall> compilerCalls,
     }
 }
 
+public readonly struct CreateFromWorkspaceResult
```
Comment (jaredpar) [inline src/Basic.CompilerLog.Util/CompilerLogUtil.cs]: Would call it `ConvertFromWorkspaceResult` for consistency. Or possibly we change it to `CreateFromBinaryLogResult`.

**Review by jaredpar — CHANGES_REQUESTED** [review]
On one hand I think this PR is on the right track if we're going to support creating from a `Workspace`. Found a few issues in the PR (see comments) but largely the code is going the right direction. 

On the other hand, the lack of fidelity for `Workspace` is a concerning. You identified that we don't have Emit data and that means it won't be as accurate as a log created from a binary log. To take this we'd need to be onboard with that. Been thinking about that and I think i'm comfortable with it so long as we can get replay / export working. Yes they won't be as high fidelity but it should still be good enough for us to put into the workflows where we consume compiler logs today.

File: src/Basic.CompilerLog.Util/CompilerLogBuilder.cs (hunk 3 above)
Comment (jaredpar) [reply src/Basic.CompilerLog.Util/CompilerLogBuilder.cs]: https://github.com/jaredpar/complog/pull/386

File: src/Basic.CompilerLog.Util/CompilerLogBuilder.cs (hunk 6 above)
Comment (jaredpar) [reply src/Basic.CompilerLog.Util/CompilerLogBuilder.cs]: That _seems_ to be the case, trying to make the contract verified + more explicit here https://github.com/jaredpar/complog/pull/387

**Comment by jaredpar** [comment]
Sorry, also meant to say in my review thank you for the work and effort you've put into this so far :smile:

File: src/Basic.CompilerLog.Util/BasicMetadataReference.cs [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+namespace Basic.CompilerLog.Util;
+
+/// <summary>
+/// A <see cref="PortableExecutableReference"/> that retains the PE image(s) it was created
+/// from. The references materialized by <see cref="CompilerLogReader"/> are stream backed:
+/// their file path is just the original file name and typically does not exist on the
+/// current machine. Retaining the images is what lets a workspace loaded from a compiler
+/// log be serialized back into a compiler log.
+/// </summary>
+internal sealed class BasicMetadataReference : PortableExecutableReference
```
Comment (jaredpar) [inline src/Basic.CompilerLog.Util/BasicMetadataReference.cs:13]: Not quite following why we need this type.

Comment (slang25) [reply src/Basic.CompilerLog.Util/BasicMetadataReference.cs:13]: This came out of adding the `WorkspaceRoundTrip` test you asked for. The references the reader materialises were built with `MetadataReference.CreateFromStream`, and once Roslyn has wrapped the bytes there's no public API to get the PE image back out — `GetMetadata()` gives you `AssemblyMetadata` rather than the image, and `MetadataImageReference` is internal. Their `FilePath` is also just the original file name, which typically doesn't exist on the machine doing the capture.

So when serialising a workspace that was loaded from a complog, `AddFromWorkspaceAsync` had no way to store the references — the round trip fell over immediately with "missing file: `Microsoft.CSharp.dll`". This type is the smallest seam I could find: it behaves the same as what `CreateFromStream` returns, but retains the images (netmodules included) so the capture path can write them straight back into the log.

I did consider materialising references to disk under `LogReaderState` instead so `FilePath` would be real, but that felt like a much bigger behavioural change for every reader. Happy to go a different way if you have something else in mind?

File: src/Basic.CompilerLog.Util/CompilerLogBuilder.cs (hunk 1 above)
Comment (slang25) [reply src/Basic.CompilerLog.Util/CompilerLogBuilder.cs]: Yeah, on reflection I agree, leaving the args empty made `replay`/`rsp`/`export` dead on arrival for workspace logs, which is a worse outcome than lossy args. I’ve brought the synthesizer back and wired it in unconditionally (it’s internal now rather than an opt-in public API), and the regression test now locks in the opposite invariant: args are always synthesized.

On noting the workspace origin during replay/export — I like the idea, but I think it needs a marker in the log metadata to know the origin at read time, so I’d rather do that as a follow-up PR than grow this one further. Happy to pick that up next if you’re keen.

File: src/Basic.CompilerLog.Util/CompilerLogBuilder.cs (hunk 2 above)
Comment (slang25) [reply src/Basic.CompilerLog.Util/CompilerLogBuilder.cs]: Done, `CreateCompilationDataPackAsync` now returns `(CompilationDataPackHash, AllProjectReferencesAdded)` as a tuple.

File: src/Basic.CompilerLog.Util/CompilerLogBuilder.cs (hunk 3 above)
Comment (slang25) [reply src/Basic.CompilerLog.Util/CompilerLogBuilder.cs]: I’ve merged main in to pick up #386 and dropped the assignment, so it’s null now.

File: src/Basic.CompilerLog.UnitTests/CompilerLogBuilderTests.cs (hunk 4 above)
Comment (slang25) [reply src/Basic.CompilerLog.UnitTests/CompilerLogBuilderTests.cs:160]: Great suggestion, this surfaced two real issues. Added `WorkspaceRoundTrip` which does exactly what you describe: `SolutionReader` → workspace → new compiler log → read the `Compilation` back → assert no error diagnostics, across every fixture log.

Making it pass needed two fixes. First, the references the reader materialises are stream backed with just a file name as the path, so capturing that workspace on another machine fell over looking for `Microsoft.CSharp.dll` on disk, there’s now an internal `BasicMetadataReference` that retains the original PE images (netmodules included) and the capture path stores those directly. Second, a log captured through the None host stores generated text but records no file-based analyzers, and `CreateNone` skipped replay when the analyzer list was empty, so `GeneratedRegex`-style projects came back with CS8795, it now replays whatever generated text is stored. No exclusions needed in the end, all 12 fixture logs round trip cleanly 🙂

File: src/Basic.CompilerLog.Util/CompilerLogBuilder.cs (hunk 9 above)
Comment (slang25) [reply src/Basic.CompilerLog.Util/CompilerLogBuilder.cs:507]: You're right, a BOM-less non-UTF-8 encoding wouldn’t have round tripped. `AddSourceText` now re-encodes those as UTF-8; anything that’s already UTF-8 or carries a preamble is written as-is.

File: src/Basic.CompilerLog.Util/CompilerLogUtil.cs (hunk 12 above)
Comment (slang25) [reply src/Basic.CompilerLog.Util/CompilerLogUtil.cs]: Done (gone with `ConvertFromWorkspaceResult`)

File: src/Basic.CompilerLog.Util/CompilerLogBuilder.cs (hunk 1 above)
Comment (slang25) [reply src/Basic.CompilerLog.Util/CompilerLogBuilder.cs]: Changed my mind, it makes sense to do the work now as it's being introduced, I've gone with a single bool to capture this to keep it simple (`IsWorkspaceLog`).

File: src/Basic.CompilerLog.Util/BasicMetadataReference.cs (hunk 13 above)
Comment (jaredpar) [reply src/Basic.CompilerLog.Util/BasicMetadataReference.cs:13]: Ah I see now.

File: src/Basic.CompilerLog.Util/Serialize/MessagePackTypes.cs [hunk 14]
```diff
@@ -317,4 +317,12 @@ public class LogInfoPack
...
     [Key(2)]
     public MSBuildDataPack? MSBuildData { get; set; }
+    /// <summary>
+    /// True when the log was created from a Roslyn <see cref="Microsoft.CodeAnalysis.Workspace"/>
+    /// rather than an actual build. Workspace logs don't include emit-time inputs so replaying
+    /// them has fidelity limits. Absent (false) on logs from older versions, which predate
+    /// workspace support and are all binary log derived.
+    /// </summary>
+    [Key(3)]
+    public bool IsWorkspaceLog { get; set; }
```
Comment (jaredpar) [inline src/Basic.CompilerLog.Util/Serialize/MessagePackTypes.cs]: Let's put this on the `CompilationInfoPack` instead. It's possible to create a log that mixes both creating from disk and workspaces. On the consumption side if say the filter on the command line only pulled out logs from the disk we wouldn't want to issue a warning. 

Yes ... a bit of an esoteric case but still possible.

File: src/Basic.CompilerLog.Util/BasicAnalyzerHost.cs [hunk 15]
```diff
@@ -130,7 +130,13 @@ BasicAnalyzerHostNone CreateNone(List<AnalyzerData> analyzers)
...
             if (analyzers.Count == 0)
             {
-                return new BasicAnalyzerHostNone();
+                // A log created from a workspace can contain generated text without recording
+                // any analyzer references (the workspace's generators are not required to be
+                // file backed). Replay whatever generated text is stored.
+                var storedSourceTexts = dataProvider.ReadAllGeneratedSourceTexts(compilerCall);
+                return storedSourceTexts.Count == 0
+                    ? new BasicAnalyzerHostNone()
+                    : new BasicAnalyzerHostNone(storedSourceTexts);
```
Comment (jaredpar) [inline src/Basic.CompilerLog.Util/BasicAnalyzerHost.cs]:
```suggestion
                var storedSourceTexts = dataProvider.ReadAllGeneratedSourceTexts(compilerCall);
                return new BasicAnalyzerHostNone(storedSourceTexts);
```

Think it can simplify to this.

File: src/Basic.CompilerLog.Util/ICompilerCallReader.cs [hunk 16]
```diff
@@ -9,6 +9,13 @@ public interface ICompilerCallReader : IDisposable
     public BasicAnalyzerKind BasicAnalyzerKind { get; }
     public LogReaderState LogReaderState { get; }
     public bool OwnsLogReaderState { get; }
+
+    /// <summary>
+    /// True when the underlying log was created from a Roslyn <see cref="Workspace"/> rather
+    /// than an actual build. Workspace logs don't include emit-time inputs (embedded resources,
+    /// Win32 resources, source link, app.config) so replay / export have fidelity limits.
+    /// </summary>
+    public bool IsWorkspaceLog { get; }
```
Comment (jaredpar) [inline src/Basic.CompilerLog.Util/ICompilerCallReader.cs]: Think this should move to `CompilerCall` once the corresponding data moves in the message pack.

**Review by jaredpar — APPROVED** [review]
Thanks for the contribution!
